include/robot_state_publisher/robot_state_publisher.hpp

```diff
@@ -34,72 +34,85 @@
 
 /* Author: Wim Meeussen */
 
-#ifndef ROBOT_STATE_PUBLISHER__ROBOT_STATE_PUBLISHER_H_
-#define ROBOT_STATE_PUBLISHER__ROBOT_STATE_PUBLISHER_H_
-
+#ifndef ROBOT_STATE_PUBLISHER__ROBOT_STATE_PUBLISHER_HPP_
+#define ROBOT_STATE_PUBLISHER__ROBOT_STATE_PUBLISHER_HPP_
+
+#include <builtin_interfaces/msg/time.hpp>
+#include <kdl/tree.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/joint_state.hpp>
+#include <std_msgs/msg/string.hpp>
+#include <tf2_ros/static_transform_broadcaster.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <urdf/model.h>
+
+#include <chrono>
 #include <map>
+#include <memory>
 #include <string>
+#include <vector>
 
-#include "kdl/frames.hpp"
-#include "kdl/segment.hpp"
-#include "kdl/tree.hpp"
-
-#include "tf2_ros/static_transform_broadcaster.h"
-#include "tf2_ros/transform_broadcaster.h"
-#include "urdf/model.h"
-
-#include "std_msgs/msg/string.hpp"
-
-#include "rclcpp/rclcpp.hpp"
+typedef std::map<std::string, urdf::JointMimicSharedPtr> MimicMap;
 
 namespace robot_state_publisher
 {
 
-class SegmentPair
+class SegmentPair final
 {
 public:
-  SegmentPair(const KDL::Segment & p_segment, const std::string & p_root, const std::string & p_tip)
+  explicit SegmentPair(
+    const KDL::Segment & p_segment,
+    const std::string & p_root,
+    const std::string & p_tip)
   : segment(p_segment), root(p_root), tip(p_tip) {}
 
   KDL::Segment segment;
-  std::string root, tip;
+  std::string root;
+  std::string tip;
 };
 
-class RobotStatePublisher
+class RobotStatePublisher final : public rclcpp::Node
```

Comment:
I don't know if anybody was actually doing this before, but this removes the ability to subclass `RobotStatePublisher`, when previously that existed.

Reply:
A quick github search didn't reveal anything other than `robot_state_publisher`'s tests using that, so it may be safe: https://github.com/ros/robot_state_publisher/blob/melodic-devel/test/test_subclass.cpp#L60-L73

Reply:
Right.  I went back and forth on this a few times.  I believe some people *were* subclassing `RobotStatePublisher` in the past, but because of the total rearrangement of the class structure in this PR, that stuff would have to be totally rewritten now.

Really, I'm hoping that the reasons people were subclassing this could go away, and that stuff could be folded back in here.  But maybe that is optimistic and I shouldn't remove the possibility here.  I'm still on the fence.

Reply:
I think that seems reasonable.

Reply:
I ended up making it sub-classable, and I removed the default destructor completely.